### PR TITLE
New hook: sphinxHook to build documentation into different formats/outputs

### PR DESCRIPTION
**Pull request by KAction**
- New hook: sphinxHook
- python3.pkgs.dropbox: use git source, not pypi tarball
- python3.pkgs.dropbox: build sphinx documentation

###### Motivation for this change

Personally, I have big overlay with offline documentation built separately,
since nixpkgs does not provide it for many packages. With this patch, we can
get proper offline upstream with little work on the side of individual
derivation author.

Price is build time, especially of pdf, and, to lesser extend, size of
build closure (texlive is quite big).

Alternative approach is to build separate
derivations with documentations. As pointed in
https://github.com/NixOS/nixpkgs/pull/136042, this has discoverity
issues, which can be solved with faking outputs in following way:

````
let
  withSphinx = opts: drv:
    let html = builtHtml (opts.sphinxRoot ? "doc/source") drv.src;
        text = builtText (opts.sphinxRoot ? "doc/source") drv.src;
        singlehtml = builtSingleHtml (opts.sphinxRoot ? "doc/source") drv.src;
    in drv // { Documentation = { inherit html text singlehtml;}; };
in
  withSphinx python3.pkgs.dropbox;
```

Extra thoughts on these approaches? Other ideas how to build offline
documentation?

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -132,6 +132,21 @@ with pkgs;
...
 
+  sphinxHook =
+    let opts = {
+          name = "sphinx-hook";
+          deps = [
+            findutils
+            python3.pkgs.sphinx
+            texinfo
+            # This is necessary to generate pdf.
+            texlive.combined.scheme-full
```
Comment (Mic92) [inline pkgs/top-level/all-packages.nix:143]: scheme-full is huge. Is there a smaller variant that works?

Comment (Mic92) [reply pkgs/top-level/all-packages.nix:143]: Afaik even `scheme-small` comes with pdflatex.

File: pkgs/build-support/setup-hooks/sphinx.sh [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+installSphinxPhase() {
+    local format o f
+    runHook preInstallSphinx
+
+    for o in $outputs ; do
+        format=$(_resolveSphinxFormat "${o}")
+        [[ -n "${format}" ]] || continue
+
+        case "$o" in
+        (pdf)
```
Comment (Mic92) [inline pkgs/build-support/setup-hooks/sphinx.sh:58]: How much do we care about pdf if we can have html?

**Review by FRidh — CHANGES_REQUESTED** [review]
- I'm of the opinion we should stick with the standard outputs we have for documentation: `doc`, `devdoc` and `man`.
- The Sphinx hook should exist for every Python interpreter, or it will cause issues when you use it with non-`python3` derivations.
- Could the hook not ensure the files are copied to the "right" place in the worktree, so the multiple-outputs hook will copy the files to the correct `doc`/`devdoc`/`man` output. This way, `sphinxHook.outputs` is not necessary anymore either; it's up to the derivation author to decide whether to use multiple outputs or not.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (KAction) [reply pkgs/top-level/all-packages.nix:143]: Yep. It also needs latexmk, cmap and tg-something (don't recall name of the package). I spent probably hour trying to cherry-pick right subset of packages, but eventually gave up and put schema-full.

File: pkgs/build-support/setup-hooks/sphinx.sh (hunk 2 above)
Comment (KAction) [reply pkgs/build-support/setup-hooks/sphinx.sh:58]: I don't know. I personally only care about singlehtml. Well, if one day sphinx finally get singletext output, I will stop caring about singlehtml either. It proves nothing, though.

**Comment by KAction** [comment]
> * I'm of the opinion we should stick with the standard outputs we have for documentation: `doc`, `devdoc` and `man`.

This way you either keep somebody unhappy by omitting their favorite format, or will force somebody to download useless format (pdf's are quite big) over poor internet connection/into small drive. Was there, not fun.

So my idea is to make things as granular as possible. Maybe we should discuss alternative approach with Nix function?

**Comment by KAction** [comment]
I decided to significantly reduce scope of the project, and now I build just html. It simplified both interface and implementation.
I also implemented suggestion of per-python versions of this hook.

@Mic92 @FRidh Please review again.

Not sure whether change to beautifulsoup4 actually belongs to this pull request, but I wanted to show that it actually works.
Current version installs documentation into `$out/share/doc/$name/html/index.html`. I am open to suggestions about better layout sheme.

File: pkgs/development/interpreters/python/sphinx-hook.nix [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{ makeSetupHook, python, findutils, sphinx }:
```
Comment (FRidh) [inline pkgs/development/interpreters/python/sphinx-hook.nix:1]: there is a folder with hooks for Python packages ./pkgs/development/interpreters/python/hooks

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/development/interpreters/python/sphinx-hook.nix (hunk 3 above)
Comment (KAction) [reply pkgs/development/interpreters/python/sphinx-hook.nix]: Fixed.

**Comment by KAction** [comment]
Ping?

**Comment by KAction** [comment]
One more revision, with even less of changes -- new `sphinxHook` derivation and two small changes for `beautifulsoup4` and `dropbox` to show that it actually works.

@FRidh Review, please.

**Comment by FRidh** [comment]
Looks good to me. 8d639f07f11d5c69ff75aaed34ec1a5f5f21aa09 commit name is not following guidelines. It is part of the package set so it should be prefixed with it.

File: pkgs/development/interpreters/python/hooks/sphinx-hook.sh [hunk 4]
```diff
@@ -0,0 +1,57 @@
+# This hook automatically finds Sphinx documentation, builds it in html format
+# and installs it.
+#
+# This hook knows about several popular locations in which subdirectory
+# documentation may be, but in very unusual cases $sphinxRoot directory can be
+# set explicitly.
+#
+# Name of the directory relative to ${doc:-$out}/share/doc is normally also
+# deduced automatically, but can be overriden with $sphinxOutdir variable.
```
Comment (SFrijters) [inline pkgs/development/interpreters/python/hooks/sphinx-hook.sh:9]: overriden -> overridden

File: pkgs/development/interpreters/python/hooks/sphinx-hook.sh [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+#
+# Sphinx build system can depend on arbitrary amount of python modules, client
+# code is responsible that all dependencies.
```
Comment (SFrijters) [inline pkgs/development/interpreters/python/hooks/sphinx-hook.sh:12]: that all dependencies -> for ensuring that all dependencies are present.

Or something like that?

**Review by SFrijters — COMMENTED** [review]
I don't know enough about the internals for an in-depth review (but I see you have an approval now already), but the code looks good to me.

**Comment by KAction** [comment]
@FRidh Fixed subject line of `sphinxHook` commit
@SFrijters Applied spelling/wording fixes.

**Pull request merged**